### PR TITLE
Ane 1164 show lernie results in scan results

### PR DESCRIPTION
**Pull request by spatten**
# Overview

[Delivers ANE-1164](https://fossa.atlassian.net/browse/ANE-1164)

Output results of keyword searches and custom license searches to stderr and the scan-summary file.

Note to reviewer: As far as I can see, there are no tests for ScanResult. I left it untested, but the test plan is pretty thorough.

## Acceptance criteria

- the results of custom-license searches are output to stderr and the scan-summary file
- The results of keyword searchs are output to stderr and the scan-summary file
- If Lernie outputs an error or warning, the error or warning should be displayed in stderr and in the scan-summary file
- If no targets are found but keyword search results are found, we should still exit with the standard "no targets found" error, but there should be an explanation that this might be okay if all you were expecting was keyword search results

## Testing plan

Scan a project with both keyword and custom-license results. Here's an example: 

[grepper.zip](https://github.com/fossas/fossa-cli/files/12528124/grepper.zip)

This project has a broken symlink in it, so we expect to see a warning emitted near the top when Lernie reports its results, and also it should say that it succeeded with 1 warning: 

```
cabal run fossa -- analyze -e http://localhost:9578 /Users/scott/fossa/license-scan-dirs/grepper --output --debug > /dev/null 
[DEBUG] Loading configuration file from "/Users/scott/fossa/license-scan-dirs/grepper/"
[DEBUG] ----------
  A task succeeded with warnings

  Warning

    walk file system: IO error for operation on /Users/scott/fossa/license-scan-dirs/grepper/s/bar: No such file or directory (os error 2): WalkFile

... lots of standard output...

* Keyword Search: succeeded with 1 warning
  ** Proprietary and Confidential match - /Users/scott/fossa/license-scan-dirs/grepper/.fossa.yml (lines 11-11)
* Custom LicenseSearch: succeeded with 1 warning
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/.fossa.yml (lines 5-5)
  ** Confidential - /Users/scott/fossa/license-scan-dirs/grepper/.fossa.yml (lines 7-7)
  ** Confidential - /Users/scott/fossa/license-scan-dirs/grepper/.fossa.yml (lines 11-11)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/three.txt (lines 1-1)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/one.txt (lines 1-1)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/one.txt (lines 3-3)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/one.txt (lines 5-5)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/one.txt (lines 7-7)

  Some projects may not appear in the summary if they were filtered during discovery.
  You can run `fossa list-targets` to see all discoverable projects.

You can also view analysis summary with warning and error messages at: "/private/var/folders/45/k76d2pgj6mlb1vpx_2gyl7z80000gn/T/fossa-analyze-scan-summary.txt"
```

The scan summary should contain the same results and the warning:

```
Scan Summary
------------
fossa-cli branch ANE-1164-show-lernie-results-in-scan-results (revision 4e237182e21d compiled with ghc-9.0)
fossa endpoint server version: N/A

2 projects scanned;  0 skipped,  2 succeeded,  0 failed,  1 analysis warning

* npm project in "/Users/scott/fossa/license-scan-dirs/grepper/": succeeded
-
* Keyword Search: succeeded with 1 warning
  ** Proprietary and Confidential match - /Users/scott/fossa/license-scan-dirs/grepper/.fossa.yml (lines 11-11)
* Custom LicenseSearch: succeeded with 1 warning
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/.fossa.yml (lines 5-5)
  ** Confidential - /Users/scott/fossa/license-scan-dirs/grepper/.fossa.yml (lines 7-7)
  ** Confidential - /Users/scott/fossa/license-scan-dirs/grepper/.fossa.yml (lines 11-11)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/three.txt (lines 1-1)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/one.txt (lines 1-1)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/one.txt (lines 3-3)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/one.txt (lines 5-5)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/one.txt (lines 7-7)
----------
* Custom-license scan & Keyword Search: succeeded with 1 warning

Warning

  walk file system: IO error for operation on /Users/scott/fossa/license-scan-dirs/grepper/s/bar: No such file or directory (os error 2): WalkFile
```

Edit the .fossa.yml so that the keyword search regex finds nothing:

```yaml
version: 3

customLicenseSearch:
  - matchCriteria: "[Pp]roprietary [Ll]icense"
    name: "Proprietary License"
  - matchCriteria: "[Cc]onfidential"
    name: "Confidential"

experimentalKeywordSearch:
  - matchCriteria: "[Pp]roprietary and [Cc]onfidential will never be found"
    name: "Proprietary and Confidential match"
```

Analyze again. The summary should explicitly say that no keyword search results were found. While we're in there, delete the empty symlink (it's the s/bar file in the directory we're scanning) so that you can see what the results look like when there are no warnings:

```
cabal run fossa -- analyze -e http://localhost:9578 /Users/scott/fossa/license-scan-dirs/grepper --output --debug > /dev/null 

* Keyword Search: succeeded
  ** No results found
* Custom LicenseSearch: succeeded
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/.fossa.yml (lines 5-5)
  ** Confidential - /Users/scott/fossa/license-scan-dirs/grepper/.fossa.yml (lines 7-7)
  ** Confidential - /Users/scott/fossa/license-scan-dirs/grepper/.fossa.yml (lines 11-11)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/three.txt (lines 1-1)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/one.txt (lines 1-1)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/one.txt (lines 3-3)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/one.txt (lines 5-5)
  ** Proprietary License - /Users/scott/fossa/license-scan-dirs/grepper/one.txt (lines 7-7)
```

Analyze a project with a bad config. For example, you can add an unclosed escape on a regex like this:

```yaml
version: 3

customLicenseSearch:
  - matchCriteria: "[Pp]roprietary [Ll]icense"
    name: "Proprietary License"
  - matchCriteria: "[Cc]onfidential\\"
    name: "Confidential"
```

When you run `fossa analyze` with that config:
- `fossa analyze` should still complete
- There should be an error saying that we failed while running Lernie, and that error should give you information on why things failed

```
cabal run fossa -- analyze -e http://localhost:9578 /Users/scott/fossa/license-scan-dirs/grepper-bad-config --output --debug > /dev/null
[DEBUG] Loading configuration file from "/Users/scott/fossa/license-scan-dirs/grepper-bad-config/"
[ERROR] ----------
  An issue occurred

  >>> Relevant errors

    Error

      Command execution failed: 

          Attempted to run the command '/private/var/folders/45/k76d2pgj6mlb1vpx_2gyl7z80000gn/T/fossa-vendor/1693946174882306/lernie --config -'
          inside the working directory '/Users/scott/fossa/fossa-cli/',
          but failed, because the command exited with code '1'.

          Often, this kind of error is caused by the project not being ready to build;
          please ensure that the project at '/Users/scott/fossa/fossa-cli/'
          builds successfully before running fossa.

          The logs for the command are listed below.
          They will likely provide guidance on how to resolve this error.

          Command standard log:
            {"type":"Error","data":{"message":"create regular expression set: '[Pp]roprietary [Ll]icense -- [Cc]onfidential\\ -- [Pp]roprietary and [Cc]onfidential': regex parse error:\n    [Cc]onfidential\\\n                   ^\nerror: incomplete escape sequence, reached end of pattern prematurely","type":"CreateRegexSet"}}


          Command error log:
            Error: create regular expression set: '[Pp]roprietary [Ll]icense -- [Cc]onfidential\ -- [Pp]roprietary and [Cc]onfidential'
            ├╴at src/config.rs:97:14
            │
            ╰─▶ regex parse error:
                    [Cc]onfidential\
                                   ^
                error: incomplete escape sequence, reached end of pattern prematurely
                ╰╴at src/config.rs:96:14


      If you believe this to be a defect, please report a bug to FOSSA support at https://support.fossa.com

      Traceback:
        - custom-license & keyword search
        - fossa-analyze


... lots of standard analysis output ...

Scan Summary
------------
fossa-cli branch ANE-1164-show-lernie-results-in-scan-results (revision 4e237182e21d compiled with ghc-9.0)
fossa endpoint server version: N/A

2 projects scanned;  0 skipped,  1 succeeded,  1 failed,  0 analysis warnings

* npm project in "/Users/scott/fossa/license-scan-dirs/grepper-bad-config/": succeeded
-
* Keyword Search: failed
* Custom LicenseSearch: failed

  Some projects may not appear in the summary if they were filtered during discovery.
  You can run `fossa list-targets` to see all discoverable projects.

You can also view analysis summary with warning and error messages at: "/private/var/folders/45/k76d2pgj6mlb1vpx_2gyl7z80000gn/T/fossa-analyze-scan-summary.txt"
```

The scan summary should also show the error:

```
Scan Summary
------------
fossa-cli branch ANE-1164-show-lernie-results-in-scan-results (revision 4e237182e21d compiled with ghc-9.0)
fossa endpoint server version: N/A

2 projects scanned;  0 skipped,  1 succeeded,  1 failed,  0 analysis warnings

* npm project in "/Users/scott/fossa/license-scan-dirs/grepper-bad-config/": succeeded
-
* Keyword Search: failed
* Custom LicenseSearch: failed
----------
* Custom-license scan & Keyword Search: failed

>>> Relevant errors

  Error

    Command execution failed: 

        Attempted to run the command '/private/var/folders/45/k76d2pgj6mlb1vpx_2gyl7z80000gn/T/fossa-vendor/1693946174882306/lernie --config -'
        inside the working directory '/Users/scott/fossa/fossa-cli/',
        but failed, because the command exited with code '1'.

        Often, this kind of error is caused by the project not being ready to build;
        please ensure that the project at '/Users/scott/fossa/fossa-cli/'
        builds successfully before running fossa.

        The logs for the command are listed below.
        They will likely provide guidance on how to resolve this error.

        Command standard log:
          {"type":"Error","data":{"message":"create regular expression set: '[Pp]roprietary [Ll]icense -- [Cc]onfidential\\ -- [Pp]roprietary and [Cc]onfidential': regex parse error:\n    [Cc]onfidential\\\n                   ^\nerror: incomplete escape sequence, reached end of pattern prematurely","type":"CreateRegexSet"}}


        Command error log:
          Error: create regular expression set: '[Pp]roprietary [Ll]icense -- [Cc]onfidential\ -- [Pp]roprietary and [Cc]onfidential'
          ├╴at src/config.rs:97:14
          │
          ╰─▶ regex parse error:
                  [Cc]onfidential\
                                 ^
              error: incomplete escape sequence, reached end of pattern prematurely
              ╰╴at src/config.rs:96:14


    If you believe this to be a defect, please report a bug to FOSSA support at https://support.fossa.com

    Traceback:
      - custom-license & keyword search
      - fossa-analyze
```

Finally, run an analysis where no targets are found, but we still find keyword search results. In this case, we should exit with a "no targets found" error, but also display helpful message that we found keyword search results and that if that's all you were expecting to find the error can be ignored.

[grepper-keyword-search-only.zip](https://github.com/fossas/fossa-cli/files/12528180/grepper-keyword-search-only.zip)

```
cabal run fossa -- analyze -e http://localhost:9578 /Users/scott/fossa/license-scan-dirs/grepper-keyword-search-only --output --debug > /dev/null
[DEBUG] Loading configuration file from "/Users/scott/fossa/license-scan-dirs/grepper-keyword-search-only/"

... lots of standard analysis output ...

Scan Summary
------------
fossa-cli branch ANE-1164-show-lernie-results-in-scan-results (revision 4e237182e21d compiled with ghc-9.0)
fossa endpoint server version: N/A

1 projects scanned;  0 skipped,  1 succeeded,  0 failed,  0 analysis warnings

-
* Keyword Search: succeeded
  ** Proprietary match - /Users/scott/fossa/license-scan-dirs/grepper-keyword-search-only/.fossa.yml (lines 5-5)
  ** Proprietary match - /Users/scott/fossa/license-scan-dirs/grepper-keyword-search-only/three.txt (lines 1-1)
  ** Proprietary match - /Users/scott/fossa/license-scan-dirs/grepper-keyword-search-only/one.txt (lines 1-1)
  ** Proprietary match - /Users/scott/fossa/license-scan-dirs/grepper-keyword-search-only/one.txt (lines 3-3)
  ** Proprietary match - /Users/scott/fossa/license-scan-dirs/grepper-keyword-search-only/one.txt (lines 5-5)
  ** Proprietary match - /Users/scott/fossa/license-scan-dirs/grepper-keyword-search-only/one.txt (lines 7-7)
* Custom LicenseSearch: succeeded
  ** No results found

  Some projects may not appear in the summary if they were filtered during discovery.
  You can run `fossa list-targets` to see all discoverable projects.

You can also view analysis summary with warning and error messages at: "/private/var/folders/45/k76d2pgj6mlb1vpx_2gyl7z80000gn/T/fossa-analyze-scan-summary.txt"
------------
[WARN] Matches to your keyword searches were found, but no other analysis targets were found.
  This will result in exiting with an error.
  This error can be safely ignored if you are only expecting keyword search results.
[ERROR] ----------
  An issue occurred

  >>> Relevant errors

    Error

      No analysis targets found in directory.

      Make sure your project is supported. See the user guide for details:
          https://github.com/fossas/fossa-cli/blob/ANE-1164-show-lernie-results-in-scan-results/docs/README.md


      Traceback:
        - fossa-analyze

  >>> Possibly-related warnings

    Warning

      >>> Relevant errors

        Error

          Could not find .git directory in the current or any parent directory

          Traceback:
            - Inferring project from git
            - Validating configuration

        Error

          Could not find executable: `svn`.
          Please ensure `svn` exists in PATH prior to running fossa.

          If you believe this to be a defect, please report a bug to FOSSA support at https://support.fossa.com

          Traceback:
            - Running command 'svn'
            - Inferring project from SVN
            - Validating configuration
```

## Risks

None that I can think of

## Metrics

Coming soon in another PR

## References

[ANE-1164](https://fossa.atlassian.net/browse/ANE-1164)

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).
~- [ ] If this PR introduced a user-visible change, I added documentation into `docs/`.~
~- [ ] If this change is externally visible, I updated `Changelog.md`. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.~
~- [ ] If I made changes to `.fossa.yml` or `fossa-deps.{json.yml}`, I updated `docs/references/files/*.schema.json`. You may also need to update these if you have added/removed new dependency type (e.g. `pip`) or analysis target type (e.g. `poetry`).~

[ANE-1164]: https://fossa.atlassian.net/browse/ANE-1164?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by csasarak — CHANGES_REQUESTED** [review]
This looks good, I added a comment about how we output error text and suggested some ideas for changing it. If they aren't workable let me know and I can approve.

File: src/App/Fossa/Analyze.hs [hunk 1]
```diff
@@ -365,15 +368,15 @@ analyze cfg = Diag.context "fossa-analyze" $ do
...
+        case (firstPartyScanResults, lernieResultsSourceUnit =<< lernieResults) of
           (Nothing, Nothing) -> Nothing
           (Just firstParty, Just lernie) -> Just $ firstParty <> lernie
           (Nothing, Just lernie) -> Just lernie
           (Just firstParty, Nothing) -> Just firstParty
   let result = buildResult includeAll additionalSourceUnits filteredProjects licenseSourceUnits
   case checkForEmptyUpload includeAll projectResults filteredProjects additionalSourceUnits licenseSourceUnits of
-    NoneDiscovered -> Diag.fatal ErrNoProjectsDiscovered
-    FilteredAll -> Diag.fatal ErrFilteredAllProjects
+    NoneDiscovered -> fatalWithKeywordSearchExplanation (maybe False (not . null . lernieResultsKeywordSearches) lernieResults) ErrNoProjectsDiscovered
```
[the remark this answers is not in the document]
Comment (spatten) [reply src/App/Fossa/Analyze.hs]: You're right, no need to output the warning and the error. I've fixed this now

File: src/App/Fossa/Analyze.hs [hunk 2]
```diff
@@ -387,6 +390,26 @@ analyze cfg = Diag.context "fossa-analyze" $ do
...
+  ) =>
+  Bool ->
+  AnalyzeError ->
+  m b
+fatalWithKeywordSearchExplanation lernieMatchesFound analyzeError = do
+  when lernieMatchesFound $
+    logWarn $
+      vsep
+        [ "Matches to your keyword searches were found, but no other analysis targets were found."
+        , "This will result in exiting with an error."
```
[the remark this answers is not in the document]
Comment (spatten) [reply src/App/Fossa/Analyze.hs]: I think it's fine to leave it with the standard error code for now. We will fix this properly in https://fossa.atlassian.net/browse/ANE-1175

**Comment by csasarak** [comment]
I like the new error output, thanks for doing that! 
Screenshot 2023-09-07 at 3 31 58 PM

**Review by csasarak — APPROVED** [review]

**Pull request merged**